exercises/src/main/scala/fpinscala/parallelism/Nonblocking.scala

```diff
@@ -123,42 +139,64 @@ object Nonblocking {
      */
     def choice[A](p: Par[Boolean])(t: Par[A], f: Par[A]): Par[A] =
       es => new Future[A] {
-        def apply(cb: A => Unit): Unit =
+        def apply(cb: A => Unit)(kErr: Throwable => Unit): Unit =
           p(es) { b =>
-            if (b) eval(es) { t(es)(cb) }
-            else eval(es) { f(es)(cb) }
-          }
+            if (b) eval(es) { t(es)(cb)(kErr) }
+            else eval(es) { f(es)(cb)(kErr) }
+          } { err => throw err }
       }
 
-    def choiceN[A](p: Par[Int])(ps: List[Par[A]]): Par[A] = ???
-
-    def choiceViaChoiceN[A](a: Par[Boolean])(ifTrue: Par[A], ifFalse: Par[A]): Par[A] =
-      ???
+    def choiceN[A](p: Par[Int])(ps: List[Par[A]]): Par[A] =
+      es => new Future[A] {
+        def apply(k: A => Unit)(kErr: Throwable => Unit): Unit =
+          p(es)(ps(_)(es)(k)(kErr))(throw _)
+      }
 
-    def choiceMap[K,V](p: Par[K])(ps: Map[K,Par[V]]): Par[V] =
-      ???
+    def choiceViaChoiceN
+      [A]
+      (a: Par[Boolean])
+      (ifTrue: Par[A], ifFalse: Par[A]):
+      Par[A] =
+      choiceN(map(a)(if (_) 0 else 1))(List(ifTrue, ifFalse))
```

Comment:
I personally try to keep it to 120 rather than 72 char, not sure why you want to keep it to 72 char, we don't print code on printers with 80 columns anymore. :)

But more to the point, there's many things that doesn't need to wrap on the next line even if you were trying to keep the line that short. For example there's no reason to wrap the parametrized type to the next line. And if you want to wrap, you should use a different indentation for the method signature than its content. For an example that still fit in a line of 72 characters:

```
    def choiceViaChoiceN[A](a: Par[Boolean])
          (ifTrue: Par[A], ifFalse: Par[A]): Par[A] =
      choiceN(map(a)(if (_) 0 else 1))(List(ifTrue, ifFalse))
```

But really, a one line signature should be perfectly good in this case: 
```
    def choiceViaChoiceN[A](a: Par[Boolean])(ifTrue: Par[A], ifFalse: Par[A]): Par[A] =
      choiceN(map(a)(if (_) 0 else 1))(List(ifTrue, ifFalse))
```